File: examples/layer/layergeode.py
# -*- Mode:Python -*-

##########################################################################
#                                                                        #
# This file is part of AVANGO.                                           #
#                                                                        #
# Copyright 1997 - 2008 Fraunhofer-Gesellschaft zur Foerderung der       #
# angewandten Forschung (FhG), Munich, Germany.                          #
#                                                                        #
# AVANGO is free software: you can redistribute it and/or modify         #
# it under the terms of the GNU Lesser General Public License as         #
# published by the Free Software Foundation, version 3.                  #
#                                                                        #
# AVANGO is distributed in the hope that it will be useful,              #
# but WITHOUT ANY WARRANTY; without even the implied warranty of         #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the           #
# GNU General Public License for more details.                           #
#                                                                        #
# You should have received a copy of the GNU Lesser General Public       #
# License along with AVANGO. If not, see <http://www.gnu.org/licenses/>. #
#                                                                        #
##########################################################################

# this script demonstrates the use of a LayerGeode. Replace LayerGeode with
# Geode to see the difference.
#
# See avango/osg/include/LayerGeode.h for information on bugs and
# missing features.

import avango.osg.simpleviewer
import math

panel1 = avango.osg.nodes.Panel(PanelColor = avango.osg.Vec4(1,0,0,0.2),
                                Width = 2,
                                Height = 1,
                                Name = "panel1",
                                )

panel2 = avango.osg.nodes.Panel(PanelColor = avango.osg.Vec4(0.8, 0.8, 0.8, 0.35),
                                Height = 1.8,
                                Name = "panel2",
                                )

panel3 = avango.osg.nodes.Panel(PanelColor = avango.osg.Vec4(0.2, 0.2, 1, 0.25),
                                Width = 1.5,
                                Height = 1.5,
                                Name = "panel3",
                                )

text = avango.osg.nodes.Text(String = "Layering is easy!",
                             Size = 0.2,
                             Alignment = 4,
                             Fontname = "VeraBI.ttf",
                             Position = avango.osg.Vec3(0,-0.2,0),
                             Color = avango.osg.Vec4(1,1,0,0.5),
                             Name = "text",
                             )

logo = avango.osg.nodes.TexturedQuad(Filename = "AvangoLogo.png",
                                     Width = 2.008,
                                     Height = 0.648,
                                     MaxAnisotropy = 16,
                                     Position = avango.osg.Vec3(-1.004, 0.42, 0),
                                     Name = "logo",
                                     )

icon = avango.osg.nodes.TexturedQuad(Filename = "delete.png",
                                     Width = 0.3,
                                     Height = 0.3,
                                     MaxAnisotropy = 16,
                                     Position = avango.osg.Vec3(-0.7, 0.12, 0),
                                     Name = "icon",
                                     )

sphere = avango.osg.nodes.Sphere(Matrix = avango.osg.make_trans_mat(0.7, 0.3, 0),
                                 Radius = 0.17,
                                 Color = avango.osg.Vec4(0.2,1,0.2,1),
                                 Name = "sphere",
                                 )

geodestate = avango.osg.nodes.StateSet(RenderingHint = 2,
                                       LightingMode = 0,
                                       BlendMode = 1,
                                       )

geode = avango.osg.nodes.LayerGeode(Drawables = [panel1,
                                                 panel2,
                                                 panel3,
                                                 logo,
                                                 text,
                                                 icon,
                                                 ],
                                    StateSet = geodestate,
                                    Name = "geode",
                                    )

trans1 = avango.osg.nodes.MatrixTransform(Matrix = avango.osg.make_trans_mat(0,0,1),
                                          Children = [geode,
                                                      sphere,
                                                      ],
                                          Name = "trans1",
                                          )

trans2 = avango.osg.nodes.MatrixTransform(Matrix = avango.osg.make_trans_mat(0,0,0),
                                          Children = [geode,
                                                      sphere,
                                                      ],
                                          Name = "trans2",
                                          )

root = avango.osg.nodes.MatrixTransform(Children = [trans1,
                                                    trans2,
                                                    ],
                                        Name = "root",
                                        Matrix = avango.osg.make_rot_mat(math.pi/2,1,0,0) * \
                                                 avango.osg.make_rot_mat(math.pi,0,0,1)
                                        )

avango.osg.simpleviewer.run(root)

File: attic/avango-osg/src/avango/osg/LayerGeode.cpp
// -*- Mode:C++ -*-

/************************************************************************\
*                                                                        *
* This file is part of Avango.                                           *
*                                                                        *
* Copyright 1997 - 2008 Fraunhofer-Gesellschaft zur Foerderung der       *
* angewandten Forschung (FhG), Munich, Germany.                          *
*                                                                        *
* Avango is free software: you can redistribute it and/or modify         *
* it under the terms of the GNU Lesser General Public License as         *
* published by the Free Software Foundation, version 3.                  *
*                                                                        *
* Avango is distributed in the hope that it will be useful,              *
* but WITHOUT ANY WARRANTY; without even the implied warranty of         *
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the           *
* GNU General Public License for more details.                           *
*                                                                        *
* You should have received a copy of the GNU Lesser General Public       *
* License along with Avango. If not, see <http://www.gnu.org/licenses/>. *
*                                                                        *
* Avango is a trademark owned by FhG.                                    *
*                                                                        *
\************************************************************************/

#include <avango/osg/LayerGeode.h>
#include <avango/Logger.h>
#include <boost/bind.hpp>
#include <osg/GL>

namespace
{
  av::Logger& logger(av::getLogger("av::osg::LayerGeode"));
}

AV_FC_DEFINE(av::osg::LayerGeode);

AV_FIELD_DEFINE(av::osg::SFLayerGeode);
AV_FIELD_DEFINE(av::osg::MFLayerGeode);

av::osg::LayerGeode::LayerGeode(::osg::Geode* osggeode) :
  Geode(osggeode),
  mLayerModeChanged(false),
  mDrawablesChanged(false)
{
  AV_FC_ADD_FIELD(LayerMode, STENCIL);
}

av::osg::LayerGeode::~LayerGeode()
{}

void
av::osg::LayerGeode::initClass()
{
  if (!isTypeInitialized())
  {
    av::osg::Node::initClass();

    AV_FC_INIT(av::osg::Node, av::osg::LayerGeode, true);

    SFLayerGeode::initClass("av::osg::SFLayerGeode", "av::Field");
    MFLayerGeode::initClass("av::osg::MFLayerGeode", "av::Field");

    sClassTypeId.setDistributable(true);
  }
}

/* virtual */ void
av::osg::LayerGeode::fieldHasChangedLocalSideEffect(const av::Field& field)
{
  Node::fieldHasChangedLocalSideEffect(field);

  if (&field == &LayerMode)
  {
    mLayerModeChanged = true;
  }
}

/* virtual */ void
av::osg::LayerGeode::evaluateLocalSideEffect()
{
  Node::evaluateLocalSideEffect();

  if (mLayerModeChanged || mDrawablesChanged)
  {
    updateDrawCallbacks();
    mLayerModeChanged = false;
    mDrawablesChanged = false;
  }
}

/* virtual */ void
av::osg::LayerGeode::setDrawablesCB(const av::osg::MFDrawable::SetValueEvent& event)
{
  Geode::setDrawablesCB(event);
  mDrawablesChanged = true;
}

void
av::osg::LayerGeode::updateDrawCallbacks()
{
  const ::osg::Geode::DrawableList& drawablelist = getOsgGeode()->getDrawableList();
  if (drawablelist.size()>1 && LayerMode.getValue()!=OFF)
  {
    drawablelist[0]->setDrawCallback(new DrawCallbackBaseLayer(this,LayerMode.getValue()));
    drawablelist[0]->setUseDisplayList(false);
    for (::osg::Geode::DrawableList::const_iterator itr=drawablelist.begin()+1; itr!=drawablelist.end(); ++itr)
    {
      (*itr)->setDrawCallback(new DrawCallbackDisableDraw());
    }
  }
  else
  {
    for (::osg::Geode::DrawableList::const_iterator itr=drawablelist.begin(); itr!=drawablelist.end(); ++itr)
    {
      (*itr)->setDrawCallback(0);
    }
  }
}

// implementation of DrawCallbackBaseLayer

av::osg::LayerGeode::DrawCallbackBaseLayer::DrawCallbackBaseLayer(av::osg::LayerGeode* geode, int mode) :
  mGeode(geode->getOsgGeode()),
  mMode(mode)
{}

/* virtual */ void
av::osg::LayerGeode::DrawCallbackBaseLayer::drawImplementation(::osg::RenderInfo& renderInfo, const ::osg::Drawable*) const
{
  switch(mMode)
  {
    case STENCIL:
      drawWithStencil(renderInfo);
      break;
    case NO_DEPTH_TEST:
      drawWithNoDepthTest(renderInfo);
      break;
    default:
      drawWithStencil(renderInfo);
  }
}

void
av::osg::LayerGeode::DrawCallbackBaseLayer::drawWithStencil(::osg::RenderInfo& renderInfo) const
{
  const ::osg::Geode::DrawableList& drawablelist = mGeode->getDrawableList();

  // get state from renderInfo
  ::osg::ref_ptr< ::osg::State> state = renderInfo.getState();

  // clear stencil buffer, enable depth and stencil test, write 1 to stencil buffer
  glPushAttrib(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT | GL_STENCIL_BUFFER_BIT);
  glClearStencil(0);
  glClear(GL_STENCIL_BUFFER_BIT);
  glEnable(GL_DEPTH_TEST);
  glEnable(GL_STENCIL_TEST);
  glStencilFunc(GL_ALWAYS, 1, 1);
  glStencilOp(GL_KEEP, GL_KEEP, GL_REPLACE);

  // because the displaylist is switched off, we need to manually apply the StateSet
  if (mGeode->getStateSet() != 0) state->pushStateSet(mGeode->getStateSet());
  state->apply();

  // apply Drawable's stateset,
  // then draw base layer in frame buffer and stencil buffer
  state->apply(drawablelist[0]->getStateSet());
  drawablelist[0]->drawImplementation(renderInfo);
  state->apply();

  // disable depth test, write colors where stencil buffer is 1, keep stencil buffer unchanged
  glDisable(GL_DEPTH_TEST);
  glStencilFunc(GL_EQUAL, 1, 1);
  glStencilOp(GL_KEEP, GL_KEEP, GL_KEEP);

  // then draw all decals in list order with their statesets applied
  for (::osg::Geode::DrawableList::const_iterator itr=drawablelist.begin()+1; itr!=drawablelist.end(); ++itr)
  {
    state->apply((*itr)->getStateSet());
    (*itr)->drawImplementation(renderInfo);
    state->apply();
  }

  if (mGeode->getStateSet() != 0) state->popStateSet();
  state->apply();

  glPopAttrib();
}

void
av::osg::LayerGeode::DrawCallbackBaseLayer::drawWithNoDepthTest(::osg::RenderInfo& renderInfo) const
{
  const ::osg::Geode::DrawableList& drawablelist = mGeode->getDrawableList();

  // get state from renderInfo
  ::osg::ref_ptr< ::osg::State> state = renderInfo.getState();

  // disable depth test
  glPushAttrib(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
  glDisable(GL_DEPTH_TEST);

  // because the displaylist is switched off, we need to manually apply the StateSet
  if (mGeode->getStateSet() != 0) state->pushStateSet(mGeode->getStateSet());
  state->apply();

  // draw all layers in list order with their statesets applied
  for (::osg::Geode::DrawableList::const_iterator itr=drawablelist.begin(); itr!=drawablelist.end(); ++itr)
  {
    state->apply((*itr)->getStateSet());
    (*itr)->drawImplementation(renderInfo);
    state->apply();
  }

  if (mGeode->getStateSet() != 0) state->popStateSet();
  state->apply();

  glPopAttrib();
}
